File: mod_src/OpenComputers-main-MC1.21.1/src/main/scala/li/cil/oc/common/Capabilities.scala
package li.cil.oc.common

import li.cil.oc.api.network.{Environment, SidedEnvironment}
import li.cil.oc.api.internal.Colored
import li.cil.oc.integration.Mods
import net.minecraft.resources.ResourceLocation
import net.minecraft.core.Direction
import net.neoforged.neoforge.capabilities.{BlockCapability, ItemCapability}
import net.neoforged.neoforge.energy.IEnergyStorage

object Capabilities {
  private def loc(path: String) = ResourceLocation.fromNamespaceAndPath(Mods.IDs.OpenComputers, path)

  val EnvironmentCapability: BlockCapability[Environment, Direction] =
    BlockCapability.createSided(loc("environment"), classOf[Environment])

  val SidedEnvironmentCapability: BlockCapability[SidedEnvironment, Direction] =
    BlockCapability.createSided(loc("sided_environment"), classOf[SidedEnvironment])

  val ColoredCapability: BlockCapability[Colored, Direction] =
    BlockCapability.createSided(loc("colored"), classOf[Colored])

  val EnergyItemCapability: ItemCapability[IEnergyStorage, java.lang.Void] =
    ItemCapability.createVoid(ResourceLocation.withDefaultNamespace("energy"), classOf[IEnergyStorage])
}

File: src/main/scala/li/cil/oc/common/capabilities/Capabilities.java
package li.cil.oc.common.capabilities;

import li.cil.oc.api.internal.Colored;
import li.cil.oc.api.network.Environment;
import li.cil.oc.api.network.SidedEnvironment;
import net.neoforged.neoforge.capabilities.BlockCapability;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;

public final class Capabilities {
    // These are defined in the Scala Capabilities object, but we keep Java stubs
    // for backwards compatibility with Java callers.
    // The actual declarations live in li.cil.oc.common.Capabilities (Scala object).

    public static BlockCapability<Colored, Direction> ColoredCapability() {
        return li.cil.oc.common.Capabilities$.MODULE$.ColoredCapability();
    }

    public static BlockCapability<Environment, Direction> EnvironmentCapability() {
        return li.cil.oc.common.Capabilities$.MODULE$.EnvironmentCapability();
    }

    public static BlockCapability<SidedEnvironment, Direction> SidedEnvironmentCapability() {
        return li.cil.oc.common.Capabilities$.MODULE$.SidedEnvironmentCapability();
    }

    private Capabilities() {
    }
}
